source/common/http/filter/extauth.cc

```diff
@@ -0,0 +1,308 @@
+                         header.key().c_str(), header.value().c_str());
+        return HeaderMap::Iterate::Continue;
+      },
+      static_cast<void*>(callbacks_));
+#endif
+}
+
+/*
+ * decodeHeaders is called at the point that the HTTP machinery handling
+ * the request has parsed the HTTP headers for this request.
+ *
+ * Our primary job here is to construct the request to the auth service
+ * and start it executing, but we also have to be sure to save a pointer
+ * to the incoming request headers in case we need to modify them in
+ * flight.
+ */
+
+FilterHeadersStatus ExtAuth::decodeHeaders(HeaderMap& headers, bool) {
+  // Remember that we have _not_ finished talking to the auth service...
+
+  auth_complete_ = false;
+
+  // ...and hang onto a pointer to the original request headers.
+  //
+  // Note that using a reference here won't work. Move semantics are the
+  // root of this issue, I think.
+
+  request_headers_ = &headers;
+
+  // Debugging.
+  dumpHeaders("decodeHeaders", request_headers_);
+
+  // OK, time to get the auth-service request set up. Create a
+  // RequestMessageImpl to hold all the details, and start it off as a
+  // copy of the incoming request's headers.
+
+  MessagePtr reqmsg(new RequestMessageImpl(HeaderMapPtr{new HeaderMapImpl(headers)}));
+
+  // We do need to tweak a couple of things. To start with, has a change
+  // to the path we hand to the auth service been configured?
+
+  if (!config_->path_prefix_.empty()) {
+    // Yes, it has. Go ahead and prepend it to the reqmsg path.
+
+    std::string path = reqmsg->headers().insertPath().value().c_str();
+
+    path = config_->path_prefix_ + path;
+
+    reqmsg->headers().insertPath().value(path);
+  }
+
+  // https://github.com/datawire/ambassador/issues/154
+  // We used to reset the Host: header to match the cluster name we're about
+  // to send the auth request to. That, however, causes trouble for anyone
+  // who wants to make auth decisions based on the host to which the client
+  // started out trying to talk to.
+  //
+  // We may need to make this configurable later, so I'm leaving this line
+  // in for reference.
+  // reqmsg->headers().insertHost().value(config_->cluster_);
+
+  // After setting up whatever headers we need to, make sure the body is
+  // correctly marked as empty.
+
+  reqmsg->headers().insertContentLength().value(uint64_t(0));
+
+  // Finally, we mark the request as being an Ambassador auth request.
+
+  reqmsg->headers().addReference(header_to_add, value_to_add.get());
+
+  // Fire the request up. When it's finished, we'll get a call to
+  // either onSuccess() or onFailure().
+
+  ENVOY_STREAM_LOG(trace, "ExtAuth contacting auth server", *callbacks_);
+
+  auth_request_ =
+      config_->cm_.httpAsyncClientForCluster(config_->cluster_)
+          .send(std::move(reqmsg), *this, Optional<std::chrono::milliseconds>(config_->timeout_));
```

Comment:
Hmm, our Optional should be able to infer the type here. I'll look into either adding that or ideally switching to abseil/folly/whatever

Reply:
Very good point. It looks like that there is already a TODO for replacing optional. Would be possible to address this in a separated PR?https://github.com/envoyproxy/envoy/blob/38d827159d5d54db20bf83734df06c90fc3fea9b/include/envoy/common/optional.h#L9

Reply:
Yeah I'll find/create an issue and get to it eventually..

Reply:
https://github.com/envoyproxy/envoy/issues/2651

Reply:
👍